dictation/dictation_context.py

```diff
@@ -0,0 +1,141 @@
+        start = max(0, self.selection.left - num_chars)
+        return self.content[start:self.selection.left]
+
+    def right_context(self, num_chars: int = DEFAULT_CONTEXT_CHARACTERS) -> str:
+        """Returns `num_chars`' worth of context to the right of the cursor"""
+        end = min(self.selection.right + num_chars, len(self.content))
+        return self.content[self.selection.right:end]
+
+
+@mod.action_class
+class ModActions:
+
+    def accessibility_dictation_enabled():
+        """Just exports `setting_accessibility_dictation` for use in other files"""
+        return setting_accessibility_dictation.get()
+    
+    def dictation_current_element() -> Element:
+        """Returns the accessibility element that should be use for dictation (i.e.
+        the current input textbox).
+        
+        This is always always the focused (current) element, however, this function
+        exists so that context can overwrite it, for applications with strange behavior.
+        """
+        return ui.focused_element()
+
+    def accessibility_adjust_context_for_application(el: Element,
+                                                     context: AccessibilityContext) -> AccessibilityContext:
+        """Sometimes the accessibility context reported by the application is wrong, but fixable
+        in predictable ways (this is most common in Electron apps).
+        
+        This method can be overwritten in those applications to do so.
+        
+        TODO(pcohen): it's a it strange to have both this and dictation_current_element;
+        possibly refactor.
+        """
+        return context
+
+    def accessibility_create_dictation_context(el: Element) -> Optional[AccessibilityContext]:
+        """Creates a `AccessibilityContext` representing the state of the input buffer
+        for dictation mode
+        """
+        if not actions.user.accessibility_dictation_enabled():
+            return None
+
+        if not el or not el.attrs:
+            # No accessibility support.
+            return None
+
+        # NOTE(pcohen): In Microsoft apps (Word, OneNote), selection will be none when the cursor
+        # is that the start of the input buffer.
+        # TODO(pcohen): this should probably be an app-specific `accessibility_adjust_context_for_application`
+        selection = el.get("AXSelectedTextRange")
+        if selection is None:
+            selection = Span(0, 0)
+            
+        context = AccessibilityContext(content=el.get("AXValue"), selection=selection)
+
+        # Support application-specific overrides:
+        context = actions.user.accessibility_adjust_context_for_application(el, context)
+
+        # If we don't appear to have any accessibility information, don't use it.
+        if context.content is None or context.selection is None:
+            return None
+
+        return context
+
+
+# TODO(pcohen): relocate this
+class Colors:
```

Comment:
Also, yuck, don't use a class for this unless you make it an Enum or something.

Reply:
Don't have an answer for this yet. I almost wish we had "utility" repo for things like this, pretty printing, and your timer class.

Reply:
Perhaps just make it an enum. Useful importing is going to have to wait until Talon has a module system, I think.

Reply:
Wouldn't be importing, but probably in action that could stylize a string with placeholders, but yeah. I will just do an enum